support/mesh/mesh_channel_core/src/deque.rs

```diff
@@ -0,0 +1,359 @@
+        Self {
+            layout: Layout::new::<T>(),
+            element_len: if size_of::<T>() == 0 {
+                align_of::<T>()
+            } else {
+                assert!(size_of::<T>() >= align_of::<T>());
+                size_of::<T>()
+            },
+            drop: if std::mem::needs_drop::<T>() {
+                Some(Self::drop_fn::<T>)
+            } else {
+                None
+            },
+        }
+    }
+
+    /// # Safety
+    /// The caller must ensure that `p` is a valid pointer to a `T`, the type
+    /// that this vtable was created with.
+    unsafe fn drop_fn<T>(p: *mut ()) {
+        // SAFETY: `p` is a valid owned pointer to a `T`.
+        unsafe {
+            drop_in_place(p.cast::<T>());
+        }
+    }
+}
+
+/// A reference to a type-erased element in the queue.
+///
+/// This is used instead of a raw pointer to ensure the queue storage is not
+/// reused while the element is still in use.
+pub struct InPlaceElement<'a>(NonNull<()>, PhantomData<&'a mut ErasedVecDeque>);
+
+impl InPlaceElement<'_> {
+    pub fn as_ptr(&self) -> *mut () {
+        self.0.as_ptr()
+    }
+}
+
+impl ErasedVecDeque {
+    /// Creates a new empty `ErasedVecDeque` with the given element vtable.
+    pub fn new(vtable: &'static ElementVtable) -> Self {
+        Self {
+            buf: NonNull::dangling(),
+            cap: if vtable.layout.size() == 0 {
+                isize::MAX as usize
+            } else {
+                0
+            },
+            head: 0,
+            len: 0,
+            vtable,
+        }
+    }
+
+    /// Returns whether the queue is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    fn offset(&self, i: usize) -> usize {
+        if self.vtable.layout.size() == 0 {
+            return 0;
+        }
+        let offset = self.head.wrapping_add(i);
+        let offset = if offset >= self.cap {
+            offset - self.cap
+        } else {
+            offset
+        };
+        debug_assert!(offset + self.vtable.layout.size() <= self.cap);
+        offset
+    }
+
+    /// # Safety
+    /// The caller must ensure that `i` is in bounds.
+    unsafe fn buf_at(&mut self, i: usize) -> InPlaceElement<'_> {
+        let offset = self.offset(i);
+        // SAFETY: `i` is a valid index into `buf`.
+        let ptr = unsafe { self.buf.add(offset).cast() };
+        InPlaceElement(ptr, PhantomData)
+    }
+
+    /// Reserves space for one element and returns a pointer to it.
+    pub fn reserve_one(&mut self) -> InPlaceElement<'_> {
+        if self.len >= self.cap {
+            self.grow();
+        }
+        // SAFETY: We just grew the buffer if necessary.
+        unsafe { self.buf_at(self.len) }
+    }
+
+    /// Commits the next element. Used after calling `reserve_one` and writing
+    /// to the buffer.
+    ///
+    /// # Safety
+    /// The caller must ensure that the element has been written to the buffer.
+    pub unsafe fn commit_one(&mut self) {
```

Comment:
this might be made marginally safer by having it accept an instance of `InPlaceElement`, to "prove" that the slot has been reserved. might require moving this method to `InPlaceElement` itself, and passing `ErasedVecDeque` to it.

ofc, I don't think there's any good way to enforce the "must ensure that the element has been written to the buffer" requirement, given the raw pointers at play... unless there's some clever use of typestates and MaybeUninit that could be useful here.

In any case though - it does seem like you need an extra clause in the SAFETY comment, reiterating that there are _two_ pre-reqs: calling `reserve_one`, and then actually doing the write via InPlaceElement

Reply:
Perhaps `reserve_one` should return `ReservedElement` that contains an actual mut ref back to the `ErasedVecDeque`, and then yes, we can put the commit on that.